src/lib/loaders/batchLoader.ts

```diff
@@ -1,55 +1,78 @@
 import DataLoader from "dataloader"
-import { chain, flatten, chunk } from "lodash"
+import { chain } from "lodash"
 import config from "config"
 import { StaticPathLoader, DynamicPathLoader } from "./api/loader_interface"
 
 const { ENABLE_RESOLVER_BATCHING } = config
 
+interface IdWithParams {
+  id: string
+  [key: string]: any
+}
+
+type SerializedParams = string[]
+
+interface BatchedParams {
+  id: string[]
+  [key: string]: any
+}
+
 /**
- * This is just used to compare options sent into the dataloader.
- * If the options are an object this stingifies the object in a way
- * that they can be grouped as similar.
+ *  This function is used by the dataloader to determine the uniqueness of keys.
+ *  Because keys can be objects, it's important to be able to differentiate between
+ *  their contents. If an object is empty except for an id field, we just use id.
+ *  Otherwise we stringify the object so it can be compared via strict comparison.
  */
-const renderParams = key => {
-  if (typeof key === "string") {
-    return ""
+export const cacheKeyFn = (key: IdWithParams): string => {
+  // If id is the *only* property of the key object
+  if (Object.keys(key).length === 1) {
+    return key.id
   }
+  return JSON.stringify(key)
+}
+
+/**
+ * This is just used to compare options sent into the dataloader.
+ */
+export const serializeParams = (key: IdWithParams) => {
   const { id, ...params } = key
   return Object.entries(params)
     .map(entry => entry.join("="))
     .sort()
     .join("&")
 }
 
-interface GroupKeysResult {
-  id: string[]
-  size: number
-  [key: string]: any
-}
-export const groupKeys = (
-  requestedKeys: string | { id }
-): GroupKeysResult[] => {
-  const grouped = chain(requestedKeys)
-    .groupBy(renderParams)
-    .values()
-    .map(values => chunk(values, 20))
-    .flatten()
-    .map((keys: string[] | { id }[]) => {
-      if (typeof keys[0] === "string") {
-        return { id: keys, size: keys.length }
-      }
-      return {
-        // @ts-ignore
-        ...keys[0],
-        id: keys.map(k => k.id),
-        size: keys.length,
+/**
+ * Takes an array of parameters and groups them by non-unique
+ * @returns a tuple of an array of group strings and an array of grouped keys
+ */
+export const groupByParams = (
+  params: IdWithParams[]
+): [SerializedParams, IdWithParams[][]] => {
+  const [serializedParams, groupedParams] = chain(params)
+    .groupBy(serializeParams)
+    .entries()
+    .thru(entries => {
+      const result: [SerializedParams, IdWithParams[][]] = [[], []]
```

Comment:
You could define this tuple type once and use it here and for the function’s return value.

Reply:
```suggestion
      const result: ParamGrouping = [[], []]
```